narwhals/_compliant/dataframe.py

```diff
@@ -364,7 +364,30 @@ def unpivot(
         value_name: str,
     ) -> Self: ...
     def with_columns(self, *exprs: CompliantExprT_contra) -> Self: ...
-    def with_row_index(self, name: str) -> Self: ...
+
+    def _with_row_index_order_by_single(self, name: str, order_by: str) -> Self:
+        plx = self.__narwhals_namespace__()
+        return self.select(
+            plx.col(order_by).rank(method="ordinal", descending=False).alias(name) - 1,
+            plx.all(),
+        )
+
+    def _with_row_index_order_by_multi(self, name: str, order_by: Sequence[str]) -> Self:
+        plx = self.__narwhals_namespace__()
+        columns = self.columns
+        const_expr = plx.lit(value=1, dtype=None).alias(name).broadcast(ExprKind.LITERAL)
+        row_index_expr = (
+            plx.col(name).cum_sum(reverse=False).over(partition_by=[], order_by=order_by)
+            - 1
+        )
```

Comment:
I moved from `rank` to `cum_sum` because it was failing in (for me...) mysterious ways